File: ReversePairs3.py
class Solution:
    def reversePairs(self, nums: list) -> int:

        def mergeSort(left, right):
            if left >= right:
                return 0
            mid = (left + right) // 2
            count = mergeSort(left, mid) + mergeSort(mid + 1, right)
            print(count)
            i, t, j = left, left, mid + 1
            cache = []
            while j <= right:
                while i <= mid and nums[i] <= nums[j] * 2:
                    i += 1
                count += mid - i + 1
                while t <= mid and nums[t] <= nums[j]:
                    cache.append(nums[t])
                    t += 1
                cache.append(nums[j])
                j += 1
            while t <= mid:
                cache.append(nums[t])
                t += 1
            for i in range(left, right + 1):
                nums[i] = cache[i - left]
            print(left, right)
            print(nums)
            return count

        return mergeSort(0, len(nums) - 1)


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.reversePairs(nums))

File: ReversePairs3.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>

using namespace std;

class Solution {
public:
    int reversePairs(vector<int>& nums) {
        return mergeSort(nums, 0, nums.size() - 1);
    }
    
    int mergeSort(vector<int>& nums, int left, int right) {
    	if(left >= right) return 0;
    	int mid = left + (right - left) / 2;
    	int res = mergeSort(nums, left, mid) + mergeSort(nums, mid + 1, right);
    	for(int i = left, j = mid + 1; i <= mid; i++) {
    		while(j <= right && nums[i] / 2.0 > nums[j]) j++;
    		res += j - (mid + 1);
		}
		
		int i = left, j = mid + 1;
		vector<int> cache;
		while(i <= mid && j <= right) {
			if(nums[i] < nums[j]) cache.push_back(nums[i++]);
			else cache.push_back(nums[j++]);
		}
		while(i <= mid) cache.push_back(nums[i++]);
		while(j <= right) cache.push_back(nums[j++]);
		
		for(i = left; i <= right; i++) nums[i] = cache[i - left];
		
//		sort(nums.begin() + left, nums.begin() + right + 1);
		return res;
	}
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	cout<<solution->reversePairs(nums);
	
	return 0;
}
